server/gfmd/auth_uid.c

```diff
@@ -262,34 +261,10 @@ auth_uid_to_global_username_kerberos(void *closure,
 	const char *auth_user_id, enum gfarm_auth_id_role *auth_user_id_rolep,
 	char **global_usernamep)
 {
-	enum gfarm_auth_id_role auth_user_id_role = *auth_user_id_rolep;
-	char *global_username;
-	struct user *u;
-	const char diag[] = "auth_uid_to_global_username_kerberos";
-
-	if (auth_user_id_role != GFARM_AUTH_ID_ROLE_USER)
-		return (GFARM_ERR_AUTHENTICATION);
-
-	giant_lock();
-	u = user_lookup_auth_id(GFARM_AUTH_USER_ID_TYPE_KERBEROS, auth_user_id);
-	giant_unlock();
-
-	if (u == NULL) {
-		/*
-		 * do not return GFARM_ERR_NO_SUCH_USER
-		 * to prevent information leak
-		 */
-		gflog_info(GFARM_MSG_UNFIXED,
-			"unknown user id <%s>, %s", auth_user_id, diag);
-		return (GFARM_ERR_AUTHENTICATION);
-	}
-	if (global_usernamep == NULL)
-		return (GFARM_ERR_NO_ERROR);
-	global_username = strdup_log(user_tenant_name(u), diag);
-	if (global_username == NULL)
-		return (GFARM_ERR_AUTHENTICATION);
-	*global_usernamep = global_username;
-	return (GFARM_ERR_NO_ERROR);
+	return (auth_uid_to_global_username_server_auth(closure,
+	  auth_user_id, gfarm_kerberos_principal_get_hostname,
+	    auth_user_id_rolep, global_usernamep,
+	    "auth_uid_to_global_username_kerberos"));
```

Comment:
This works for the GFARM_AUTH_ID_ROLE_SPOOL_HOST case and the GFARM_AUTH_ID_ROLE_METADATA_HOST case.
But for the GFARM_AUTH_ID_ROLE_USER case, auth_uid_to_global_username_kerberos() should do auth_user_id -> gfarm global user id mapping using the user_id hash table, It does not do it for now.

Please modify as follows:
* Please create a new function user_lookup_by_kerberos_principal() that converts a Kerberos principal name to a Gfarm global user name using the user_id hash table.
* Please add a new input parameter struct user *(*user_lookup_by_auth_user_id)(const char *auth_user_id) to auth_uid_to_global_username_server_auth()
* In the tls_client_certificate case, please pass the user_lookup_gsi_dn function as the user_lookup_by_auth_user_id parameter.
* In the Kerberos case, please pass a new user_lookup_by_kerberos_principal function as the user_lookup_by_auth_user_id parameter.